ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/tasks/DownloadHeaderSequenceTask.java

```diff
@@ -179,73 +204,126 @@ private CompletableFuture<PeerTaskResult<List<BlockHeader>>> downloadHeaders(
-                    if (invalidHeader == null) {
-                      final String msg =
-                          String.format(
-                              "Received misordered blocks. Missing child of %s",
-                              header.toLogString());
-                      exception = InvalidBlockException.create(msg);
-                    } else {
-                      final String errorMsg =
-                          headerInvalid
-                              ? "Header failed validation"
-                              : "Out-of-range header received from peer";
-                      exception = InvalidBlockException.fromInvalidBlock(errorMsg, invalidHeader);
-                    }
-                    future.completeExceptionally(exception);
-                  });
+  private CompletableFuture<List<BlockHeader>> processHeadersUsingPeerTask(
+      final PeerTaskExecutorResult<List<BlockHeader>> headersResult) {
+    final List<BlockHeader> blockHeaders =
+        headersResult
+            .result()
+            .orElseThrow(
+                () -> new RuntimeException("Expected blockHeaders in PeerTaskExecutorResult"));
+    final EthPeer ethPeer =
+        headersResult
+            .ethPeer()
+            .orElseThrow(() -> new RuntimeException("Expected a peer in PeerTaskExecutorResult"));
+    return processHeaders(blockHeaders, ethPeer);
+  }
 
-              return future;
-            }
-            headers[headerIndex] = header;
-            lastFilledHeaderIndex = headerIndex;
-            child = header;
-          }
-          future.complete(asList(headers).subList(lastFilledHeaderIndex, previousHeaderIndex));
-          return future;
-        });
+  private CompletableFuture<List<BlockHeader>> processHeaders(
+      final List<BlockHeader> blockHeaders, final EthPeer ethPeer) {
+    final CompletableFuture<List<BlockHeader>> future = new CompletableFuture<>();
+    BlockHeader child = null;
+    boolean firstSkipped = false;
+    final int previousHeaderIndex = lastFilledHeaderIndex;
+    for (final BlockHeader header : blockHeaders) {
+      final int headerIndex =
+          Ints.checkedCast(segmentLength - (referenceHeader.getNumber() - header.getNumber()));
+      if (!firstSkipped) {
+        // Skip over reference header
+        firstSkipped = true;
+        continue;
+      }
+      if (child == null) {
+        child = (headerIndex == segmentLength - 1) ? referenceHeader : headers[headerIndex + 1];
+      }
+
+      final boolean foundChild = child != null;
+      final boolean headerInRange = checkHeaderInRange(header);
+      final boolean headerInvalid = foundChild && !validateHeader(child, header);
+      if (!headerInRange || !foundChild || headerInvalid) {
+        final BlockHeader invalidHeader = child;
+        final CompletableFuture<?> badBlockHandled =
+            headerInvalid
+                ? markBadBlock(invalidHeader, ethPeer)
```

Comment:
not so sure about the bad block handling, but I think nothing has changed there ...